ios/brave-ios/Sources/Brave/WebFilters/FilterListResourceDownloader.swift

```diff
@@ -50,83 +44,51 @@ public actor FilterListResourceDownloader {
 
         await AdBlockGroupsManager.shared.didUpdateResourcesComponent(folderURL: folderURL)
         await FilterListCustomURLDownloader.shared.startIfNeeded()
-
-        if !FilterListStorage.shared.filterLists.isEmpty {
-          await registerAllFilterListsIfNeeded(with: adBlockService)
-        }
+        await subscribeToFilterListChanges(with: adBlockService)
       }
     }
+  }
 
-    Task { @MainActor in
-      for await filterListEntries in adBlockService.filterListCatalogComponentStream() {
-        FilterListStorage.shared.loadFilterLists(from: filterListEntries)
-        ContentBlockerManager.log.debug("Loaded filter list catalog")
-
-        if AdBlockGroupsManager.shared.resourcesInfo != nil {
-          await registerAllFilterListsIfNeeded(with: adBlockService)
+  /// Subscribe to filter list changes so we keep the engines up to date
+  private func subscribeToFilterListChanges(with adBlockService: AdblockService) async {
+    guard !subscribedToFilterListChange else { return }
+    self.subscribedToFilterListChange = true
+    filterListSubscription = FilterListStorage.shared.$filterLists
+      .receive(on: DispatchQueue.main)
+      .sink { [weak self] _ in
+        Task { @MainActor in
```

Comment:
Why `Task { @MainActor }` when you already have `receive(on: .main)`?

Reply:
It's because specifying receive(on: .main) does not tell the next callback that it's now on a MainActor isolation. Therefore im forced to put the Task. However I can probably remove the `receive(on: .main)` but I definitely can't remove the Task.

Reply:
I don't think so. The task should not be necessary here: https://developer.apple.com/documentation/combine/publisher/receive(on:options:)
```swift
pub.sink {
    DispatchQueue.main.async {
        // Do something.
    }
}
```
is the same as:
```swift
pub.receive(on: DispatchQueue.main)
.sink {
    // Do something.
}
```

Reply:
I don't think so. https://developer.apple.com/documentation/combine/publisher/receive(on:options:)
```swift
pub.sink {
    Task { @MainActor in
        // Do something.
    }
}
```
is the same as:
```swift
pub.receive(on: DispatchQueue.main)
.sink {
    // Do something.
}
```